### PR TITLE
FROM ONLY <table> clause

### PR DESCRIPTION
**Pull request by Ploppz**
Compile fail test don't pass yet due to https://github.com/dtolnay/trybuild/issues/140 but I would like to already request feedback on, for example, the placement of my implementation and how to expose it in the API.

### CODE REVIEW

File: diesel/src/macros/mod.rs [hunk 1]
```diff
@@ -810,6 +814,9 @@ macro_rules! __diesel_table_impl {
                 }
             }
 
+            // impl $crate::expression::AppearsOnTable<$crate::query_builder::Only<table>> for <table as $crate::query_source::QuerySource>::DefaultSelection {}
```
Comment (Ploppz) [inline diesel/src/macros/mod.rs:817]: This resulted in an impl conflict error, but apparently was not needed after all.

File: diesel/src/query_builder/only_clause.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+/// Represents a query with an `ONLY` clause.
+#[derive(Debug, QueryId)]
+pub struct Only<T> {
+    query: T,
+}
+
+/// Add `ONLY` before a query source.
+pub trait OnlyDsl: AsQuery + Sized {
```
Comment (weiznich) [inline diesel/src/query_builder/only_clause.rs:19]: The dsl should be moved to the querydsl module (or the corresponding submodule of the pg module)

File: diesel/src/query_builder/only_clause.rs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    query_source::Table,
+};
+
+/// Represents a query with an `ONLY` clause.
+#[derive(Debug, QueryId)]
+pub struct Only<T> {
+    query: T,
+}
+
+/// Add `ONLY` before a query source.
```
Comment (weiznich) [inline diesel/src/query_builder/only_clause.rs:18]: This is the main entry point for using `from only ...` clauses. I feel the documentation here could be a bit more detailed.

File: diesel/src/macros/mod.rs (hunk 1 above)
Comment (weiznich) [reply diesel/src/macros/mod.rs]: Yes this is not needed if we implement the corresponding traits for columns, as this is then implied by a more general impl.

**Review by weiznich — COMMENTED** [review]
Thanks for submitting this pr :+1:
I've added a few comments on how to improve the general structure of the changed code.

File: diesel/src/query_builder/only_clause.rs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    fn walk_ast(&self, mut out: AstPass<DB>) -> QueryResult<()> {
+        out.push_sql("ONLY ");
+        self.query.walk_ast(out.reborrow())?;
+        Ok(())
+    }
+}
+
+impl<T> QueryDsl for Only<T> {}
+
+impl<T, S> SelectDsl<S> for Only<T>
```
[the remark this answers is not in the document]
Comment (Ploppz) [reply diesel/src/query_builder/only_clause.rs:55]: What do you mean?

File: diesel/src/query_builder/only_clause.rs (hunk 2 above)
Comment (Ploppz) [reply diesel/src/query_builder/only_clause.rs:19]: I can't find any dsl module in `src/pg`

Comment (weiznich) [reply diesel/src/query_builder/only_clause.rs:19]: I would likely add it [here](https://github.com/diesel-rs/diesel/tree/master/diesel/src/pg/expression/extensions) as additional sub module + reexport it from `diesel::query_dsl`

File: diesel/src/query_builder/only_clause.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (weiznich) [reply diesel/src/query_builder/only_clause.rs:55]: The current implementation only allows users to call `.select()` on any table you've called `.only()` on. To allow calling other `QueryDsl` methods, like for example filter, we need to implement more traits. Basically every trait reexported [here](https://github.com/diesel-rs/diesel/blob/master/diesel/src/query_dsl/mod.rs#L63-L78) needs to be implemented.

File: diesel/src/query_builder/only_clause.rs (hunk 2 above)
Comment (Ploppz) [reply diesel/src/query_builder/only_clause.rs:19]: I guess it could go directly into the `pg` module?

**Comment by Ploppz** [comment]
I updated the commit. Is the structure good now? I put `Only` and `OnlyDsl` in different places, not sure about that though.

**Comment by weiznich** [comment]
Thanks for the update :+1:

The general structure looks much nicer now. I left a comment on a minor improvement and a comment about how to fix the compilation errors. Beside of that I would like to see the following things implemented/added

- [ ] A bit more documentation on the `OnlyDsl` trait
- [ ] A changelog entry
- [ ] Adding the missing query dsl impls.

File: diesel/src/macros/mod.rs [hunk 5]
```diff
@@ -135,6 +135,10 @@ macro_rules! __diesel_column {
             }
         }
 
+
+        impl $crate::expression::AppearsOnTable<$crate::pg::Only<$table>> for $column_name {}
```
[the remark this answers is not in the document]
Comment (Ploppz) [reply diesel/src/macros/mod.rs:139]: ah indeed I ran into problems when I tried that `#[cfg(feature = "postgres_backend")]`. Thanks, will do that

**Comment by Ploppz** [comment]
Does it look good now?

File: diesel_compile_tests/tests/fail/only_only_on_table.rs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+        id -> Int8,
+    }
+}
+
+fn main() {
+    let mut connection = PgConnection::establish("").unwrap();
+
+    foo::table.select(foo::id).only();
+    foo::table.select(foo::id).filter(foo::id.eq(1)).only();
+    foo::table.select(foo::id.only());
```
Comment (weiznich) [inline diesel_compile_tests/tests/fail/only_only_on_table.rs:17]: I would like to see here the following additional checks:

```rust
let mut conn = SqliteConnection::establish("").unwrap();

foo::table.only().load(connection).unwrap();

let mut conn = MysqlConnection::establish("").unwrap();

foo::table.only().load(connection).unwrap();
```

**Review by weiznich — APPROVED** [review]
This looks good now, beside the three minor things mentioned in the review comments.
Thanks for your work here :+1:

File: diesel/src/pg/expression/extensions/only_dsl.rs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+///     .select(count(users::id))
+///     .first::<i64>(connection)
+///     .unwrap();
+///
+/// ```
+/// Selects the number of entries in the `users` table excluding any rows found in inherited
+/// tables.
+pub trait OnlyDsl: AsQuery + Sized {
+    /// See the trait-level docs.
+    fn only(self) -> SelectStatement<Only<Self>> {
```
[the remark this answers is not in the document]
Comment (weiznich) [reply diesel/src/pg/expression/extensions/only_dsl.rs:26]:
```suggestion
    fn only(self) -> crate::dsl::SelectFromOnly<Self> {
```

**Review by weiznich — APPROVED** [review]

**Pull request merged**